google-cloud-spanner/src/test/java/com/google/cloud/spanner/AsyncResultSetImplTest.java

```diff
@@ -301,6 +305,72 @@ public void pauseResume() throws InterruptedException {
     }
   }
 
+  @Test
+  public void testCallbackIsNotCalledWhilePaused() throws InterruptedException, ExecutionException {
+    Executor executor = Executors.newSingleThreadExecutor();
+    final int simulatedRows = 100;
+    ResultSet delegate = mock(ResultSet.class);
+    when(delegate.next())
+        .thenAnswer(
+            new Answer<Boolean>() {
+              int row = 0;
+
+              @Override
+              public Boolean answer(InvocationOnMock invocation) throws Throwable {
+                row++;
+                if (row > simulatedRows) {
+                  return false;
+                }
+                return true;
+              }
+            });
+    when(delegate.getCurrentRowAsStruct()).thenReturn(mock(Struct.class));
+    final AtomicInteger callbackCounter = new AtomicInteger();
+    final BlockingDeque<Object> queue = new LinkedBlockingDeque<>(1);
+    final AtomicBoolean paused = new AtomicBoolean();
+    try (AsyncResultSetImpl rs =
+        new AsyncResultSetImpl(simpleProvider, delegate, AsyncResultSetImpl.DEFAULT_BUFFER_SIZE)) {
+      ApiFuture<Void> callbackResult =
+          rs.setCallback(
+              executor,
+              resultSet -> {
+                assertFalse(paused.get());
+                callbackCounter.incrementAndGet();
+                try {
+                  while (true) {
+                    switch (resultSet.tryNext()) {
+                      case OK:
+                        paused.set(true);
+                        queue.put(new Object());
+                        return CallbackResponse.PAUSE;
+                      case DONE:
+                        return CallbackResponse.DONE;
+                      case NOT_READY:
+                        return CallbackResponse.CONTINUE;
+                    }
+                  }
+                } catch (InterruptedException e) {
+                  throw SpannerExceptionFactory.propagateInterrupt(e);
+                }
+              });
+      int rowCounter = 0;
+      while (!callbackResult.isDone()) {
+        Object o = queue.poll(1L, TimeUnit.MILLISECONDS);
+        if (o != null) {
+          rowCounter++;
+        }
+        Thread.yield();
```

Comment:
If I understood this correctly, we could also sleep after this for a while (although I am not super happy with sleeps in tests) in order to guarantee that the callback does not run until resume is called.

Reply:
(minor correction, we don't need to yield if we sleep)

Reply:
Yes, that's correct. I tried it locally with a number of different sleep times, and it succeeded in all cases. I then changed it to `Thread.yield()` as I'm also not really a fan of `Thread.sleep()` in a test case if we can avoid it.

Reply:
Ah great, that would be the approach I would take as well. LGTM